proposals/lip-0054.md

```diff
@@ -301,104 +303,141 @@ messageRecoveryParams = {
+            raise Exception("Cross-chain message is not pending.")
+
+    # Check the validity of the CCMs to be recovered
+    deserializedCCMs = [decode(crossChainMessageSchema, ccmBytes) for ccmBytes in trs.params.crossChainMessages]
+    for ccm in deserializedCCMs:
+        if ccm.status != CCM_STATUS_OK:
+            raise Exception("Cross-chain message status is not valid.")
+
+    # Check the inclusion proof against the sidechain outbox root
+    proof = { size: terminatedOutboxAccount(trs.params.chainID).outboxSize,
+        idxs: trs.params.idxs,
+        siblingHashes: trs.params.siblingHashes}
+
+    if RMTVerify(
+        [SHA-256(ccmBytes) for ccmBytes in trs.params.crossChainMessages],
+        proof,
+        terminatedOutboxAccount(trs.params.chainID).outboxRoot
+        ) == False:
+        raise Exception("Message recovery proof of inclusion is not valid.")
 ```
 
-#### Message Recovery Command Execution
-
-Processing a transaction `trs` with module ID `MODULE_ID_INTEROPERABILITY` and command ID `COMMAND_ID_MESSAGE_RECOVERY` implies the following logic:
+#### Execution
 
 ```python
-let trsSenderAddress be the address of the trs.senderPublicKey
-
-# Set CCM status to recovered and assign fee to trs sender
-updatedCCMs = []
-for ccm in deserializedCCMs:
-
-    for each module mdl for which beforeRecoverCCM exists:
-        mdl.beforeRecoverCCM(trsSenderAddress, ccm)
-        if the previous call fails:
-            execution fails and trs is invalid
-    ccm.status = CCM_STATUS_RECOVERED
-    ccm.fee = 0
-    push serialized(ccm) to updatedCCMs # CCM is serialized again
-
-# Update sidechain outbox root
-proof = { size: terminatedOutboxAccount(trs.params.chainID).outboxSize,
-    idxs: trs.params.idxs,
-    siblingHashes: trs.params.siblingHashes}
-
-terminatedOutboxAccount(trs.params.chainID).outboxRoot = RMTCalculateRoot([SHA-256(ccmData) for CCMData in updatedCCMs], proof)
-
-# Process recovery
-for ccm in deserializedCCMs:
-    swap ccm.sendingChainID and ccm.receivingChainID
-
-    # On sidechain, this case is always triggered and the message logic should be supported
-    if ccm.receivingChainID == ownChainAccount.ID:
-        if (ccm.moduleID, ccm.crossChainCommandID) is supported:
-            call the logic associated with (ccm.moduleID, ccm.crossChainCommandID) on ccm
-    elif ownChainAccount.ID == MAINCHAIN_ID:
-        # Only send messages to active and live chains
-        if (chainAccount(ccm.receivingChainID) exists
-            and chainAccount(ccm.receivingChainID).status == CHAIN_ACTIVE
+def execute(trs: Transaction) -> None:
+    senderAddress = SHA-256(trs.senderPublicKey)[:ADDRESS_LENGTH]
+    # Set CCM status to recovered and assign fee to trs sender
+    crossChainMessages = [decode(crossChainMessageSchema, ccmBytes) for ccmBytes in trs.params.crossChainMessages]
+    updatedCCMs = []
+    for ccm in crossChainMessages:
+        token.assignMessageFee(senderAddress, ccm)
```

Comment:
I haven't found an `assignMessageFee` function in the Token module.

Reply:
it is done in `beforeExecuteCCM`

Reply:
I fixed the function call. However, I will revisit the whole processing of recoveries after LIP53 is finalized as I think it should mirror the CCU execution stages